File: app/src/main/java/com/boardgamegeek/sorter/RatingSorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;

import com.boardgamegeek.R;

import java.text.DecimalFormat;

public abstract class RatingSorter extends CollectionSorter {
	private final String defaultValue;

	public RatingSorter(@NonNull Context context) {
		super(context);
		defaultValue = context.getString(R.string.text_unknown);
	}

	@Override
	protected boolean isSortDescending() {
		return true;
	}

	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		return getInfo(cursor, null);
	}

	@Override
	public String getDisplayInfo(@NonNull Cursor cursor) {
		return getInfo(cursor, getDisplayFormat());
	}

	protected String getInfo(@NonNull Cursor cursor, DecimalFormat format) {
		return getDoubleAsString(cursor, getSortColumn(), defaultValue, true, format);
	}

	protected abstract DecimalFormat getDisplayFormat();
}

File: app/src/main/java/com/boardgamegeek/sorter/RatingSorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import com.boardgamegeek.R
import com.boardgamegeek.model.CollectionItem
import com.boardgamegeek.extensions.asBoundedRating
import java.text.DecimalFormat

abstract class RatingSorter(context: Context) : CollectionSorter(context) {
    private val defaultValue: String = context.getString(R.string.text_unknown)

    protected abstract val displayFormat: DecimalFormat

    override fun getHeaderText(item: CollectionItem): String {
        val rating = getRating(item)
        return if (rating == 0.0) defaultValue else rating.asBoundedRating(context, DecimalFormat("#0.0"), R.string.unrated_abbr)
    }

    override fun getRatingText(item: CollectionItem) = getRating(item).asBoundedRating(context, displayFormat, R.string.unrated_abbr)
}
